core/trino-spi/src/test/java/io/trino/spi/block/BenchmarkVariableWidthBlock.java

```diff
@@ -0,0 +1,214 @@
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.openjdk.jmh.annotations.Mode.AverageTime;
+import static org.openjdk.jmh.annotations.Scope.Thread;
+
+@OutputTimeUnit(MICROSECONDS)
+@Fork(2)
+@Warmup(iterations = 5, time = 1000, timeUnit = MILLISECONDS)
+@Measurement(iterations = 10, time = 1000, timeUnit = MILLISECONDS)
+@BenchmarkMode(AverageTime)
+public class BenchmarkVariableWidthBlock
+{
+    private static final int SEED = 831;
+    private static final int POSITIONS = 8096;
+    private static final double NULLS_CHANCE = 0.2;
+
+    @Benchmark
+    public Block copyPositions(BenchmarkData data)
+    {
+        int[] positionIds = data.getPositionsIds();
+        return data.getVariableWidthBlock().copyPositions(positionIds, 0, positionIds.length);
+    }
+
+    @State(Thread)
+    public static class BenchmarkData
+    {
+        @Param({"200", "1000", "8000"})
+        private int selectedPositionsCount;
+
+        @Param({"false", "true"})
+        private boolean nullsAllowed;
+
+        @Param({
+                "GROUPED",
+                "SEQUENCE",
+                "RANDOM",
+        })
+        private SelectedPositions selectedPositions;
+
+        private int[] positionsIds;
+        private Block variableWidthBlock;
+
+        public BenchmarkData(int selectedPositionsCount, boolean nullsAllowed, SelectedPositions selectedPositions)
+        {
+            this.selectedPositionsCount = selectedPositionsCount;
+            this.nullsAllowed = nullsAllowed;
+            this.selectedPositions = selectedPositions;
+        }
+
+        public BenchmarkData()
+        {
+            this(1000, false, SelectedPositions.SEQUENCE);
+        }
+
+        @Setup
+        public void setup()
+        {
+            positionsIds = selectedPositions.generateIds(selectedPositionsCount);
+            Slice[] slices = generateValues(nullsAllowed);
+            variableWidthBlock = createBlockBuilderWithValues(slices).build();
+        }
+
+        private static Slice[] generateValues(boolean nullsAllowed)
+        {
+            Random random = new Random(SEED);
+            Slice[] generatedValues = new Slice[POSITIONS];
+            for (int position = 0; position < POSITIONS; position++) {
+                if (nullsAllowed && isTrue(random, NULLS_CHANCE)) {
+                    generatedValues[position] = null;
+                }
+                else {
+                    int length = random.nextInt(380) + 20;
+                    byte[] buffer = new byte[length];
+                    random.nextBytes(buffer);
+                    generatedValues[position] = Slices.wrappedBuffer(buffer);
+                }
+            }
+            return generatedValues;
+        }
+
+        private static boolean isTrue(Random random, double chance)
+        {
+            double value = 0;
+            // chance has to be 0 to 1 exclusive.
+            while (value == 0) {
```

Comment:
Why 0 is bad?

Reply:
Not sure, I just copied this one from `BenchmarkDataGenerator#randomNullChance`
@skrzypo987 do you know why ?

Reply:
There is a `<` at the end, so if random picks 0 it will return null for null chance of 0.
We basically want to pick random out of [0,1>, but `nextBoolean` returns <0,1>

Reply:
I would simplify it